ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/LeakyBucket.hs

```diff
@@ -191,42 +271,67 @@ init ::
+  newTVarIO $
+    State
+      { time,
+        level = capacity,
+        paused = False,
+        configGeneration = 0,
+        config = config
+      }
 
 -- | Monadic action that calls 'threadDelay' until the bucket is empty, then
--- returns @()@. It receives the 'ThreadId' argument of the action's thread,
--- which it uses to throw exceptions at it; it returns a monadic action that can
--- be used to interrupt the thread from the outside.
+-- runs the 'onEmpty' action and terminates.
 leak ::
   (MonadDelay m, MonadCatch m, MonadFork m, MonadAsync m) =>
-  Bucket m ->
+  -- | A variable indicating whether the thread should run (when it is filled)
+  -- or not (otherwise). The integer it carries only helps in differentiating
+  -- between starts and restarts. 'leak' does not modify this variable.
+  StrictTMVar m Int ->
+  -- | The 'ThreadId' of the action's thread, which is used to throw exceptions
+  -- at it.
   ThreadId m ->
-  m (Maybe (m ()))
-leak bucket actionThreadId = do
-  State {config = Config {rate}} <- snapshot bucket
-  if rate <= 0
-    then pure Nothing
-    else do
-      a <- async go
-      pure $ Just $! uninterruptibleCancel a
+  Bucket m ->
+  m ()
+leak runThreadVar actionThreadId bucket = go
   where
     go = do
-      State {level, config = Config {rate, onEmpty}} <- snapshot bucket
+      -- Block until we are allowed to run. Do not modify the TMVar.
+      oldRunThread <- atomically $ readTMVar runThreadVar
+      -- NOTE: It is tempting to group this @atomically@ and
+      -- @atomicallyWithMonotonicTime@ into one; however, because the former is
+      -- blocking, the latter could get a _very_ inaccurate time, which we
+      -- cannot afford.
+      State {level, configGeneration = oldConfigGeneration, config = Config {rate, onEmpty}} <-
+        atomicallyWithMonotonicTime $ snapshot bucket
       let timeToWait = secondsRationalToDiffTime (level / rate)
       -- NOTE: It is possible that @timeToWait == 0@ while @level > 0@ when @level@
       -- is so tiny that @level / rate@ rounds down to 0 picoseconds. In that case,
       -- it is safe to assume that it is just zero.
       if level <= 0 || timeToWait == 0
-        then handle (\(e :: SomeException) -> throwTo actionThreadId e) onEmpty
-        else threadDelay timeToWait >> go
+        then do
+          handle (\(e :: SomeException) -> throwTo actionThreadId e) onEmpty
+          -- We have run the action on empty, there is nothing left to do,
+          -- unless someone changes the configuration.
+          void $ atomically $ blockUntilChanged configGeneration oldConfigGeneration $ readTVar bucket
+          go
+        else do
+          -- Wait for the bucket to empty, or for the thread to be stopped or
+          -- restarted.
+          race_
+            (threadDelay timeToWait)
+            (void $ atomically $ blockUntilChanged id (Just oldRunThread) $ tryReadTMVar runThreadVar)
```

Comment:
Good point! Feels to me like we should go for [`timeout`](https://hackage.haskell.org/package/base-4.18.1.0/docs/System-Timeout.html#v:timeout) instead, though? (In `io-classes`'s `MonadTimer`.) Which would give:
```haskell
void $ timeout timeToWait $
  blockUntilChanged ...
```

Reply:
That would likely also have slightly more overhead as it also involves a `forkIO` for sending an async exception to the main thread on timeout.

Reply:
Good point. `io-classes` doesn't seem to give me the instance for `Alternative (STM m)` though; did I miss something or is it not there?

Reply:
Oh right, surprising that it doesn't provide it (might create a PR adding it). You can `orElse`, that's what `<|>` is implemented with.

Reply:
Hum, this requires propagating a `MonadTimer m` constraint everywhere including in the ChainSync client and everything that uses it. Apparently that is not in IOLike?

Reply:
I think it would be fine to add it to `IOLike`, it already includes `MonadDelay` which contains `threadDelay`, and one can implement `registerDelay` and `timeout` already using that, just less efficiently.

Reply:
(If that requires too much yak shaving, feel free to keep eg the `race` version + a TODO in a comment.)

Reply:
It was not that bad, actually. Done in 2b0fdd3215e08df4a68e75b4aad5a4aac640dbcd. Does this look good to you?